File: Algorithms/QuickSort/quick_sort.cpp
#include <bits/stdc++.h>
using namespace std;

/*
left side <= pivot and right side > pivot
*/

int partition(int A[], int start, int end);

void quick_sort(int A[], int start, int end);

int main(){
    // int A[] = {7, 6, 5, 4, 3, 2, 1, 0};
    int sizeofarray;
    cout << "Enter the array length: ";
    cin >> sizeofarray;
    
    int A[sizeofarray];

    for (int i = 0; i < sizeofarray; i++){
        cin >> A[i];
    }
    quick_sort(A, 0, sizeofarray);
    for (int i = 0; i < sizeofarray; i++) cout << A[i] << " ";
    return 0;
}

int partition(int A[], int start, int end){
    int pivot = A[end];              // set last index as pivot
    int partitionIndex = start;     // set partition index as start initially

    for (int i = start; i < end; i++){
        if(A[i] <= pivot){
            swap(A[i], A[partitionIndex]);  // swap if element is lesser than pivot
            partitionIndex++;
        }
    }
    swap(A[partitionIndex], A[end]);        // swap pivot with element at partition index in middle
    return partitionIndex;
}

void quick_sort(int A[], int start, int end){
    if (start < end){
        int partitionIndex = partition(A, start, end);  // calling partition
        quick_sort(A, start, partitionIndex-1);
        quick_sort(A, partitionIndex+1, end);
    }
}

File: Algorithms/QuickSort/quick_sort.py
def partition(arr, start, end):
    pivot = arr[start]            # set the pivot as initially its the first index
    low = start + 1 
    high = end

    while True:
        """
        If the current value we're looking at is larger than the pivot
        it's in the right place(right side of pivot) and we can move left to the next element
        we also need to make sure we haven't surpassed the low pointer, since that
        indicates we have already moved all the elements to their correct side of the pivot
        """
        while low <= high and arr[high] >= pivot:
            high = high - 1
        
        # opposite process of the one above
        while low <= high and arr[low] <= pivot:
            low = low + 1

        # we either found a value for both high and low that is out of order
        # or low is higher than high, in which case we exit the loop
        if low <= high:
            arr[low], arr[high] = arr[high], arr[low]
        else:
            break
    arr[start], arr[high] = arr[high], arr[start]
    return high


def quick_sort(arr, start, end):
    if start >= end:
        return
    # partitioning index that is arr[p] is now
    partition_index = partition(arr, start, end)

    # separately sort elemenets before
    # partition and after partition
    quick_sort(arr, start, partition_index-1)   # before the pivot for first sub-array
    quick_sort(arr, partition_index+1, end)     # after the pivot values (2nd subarray)


if __name__ == "__main__":
    arr = [5, 6, 7, 3, 2, 4]
    arr_len = len(arr)

    quick_sort(arr, 0, arr_len-1)
    print(arr)